Support cumulative, delta, and pass-through exporters

jmacd opened the pull request:
With these changes, Cumulative and Delta-oriented Exporters are able to request the correct kind of aggregation. This is the final step in a re-do of #799, an effort which included #808, #812, and #835.

Resolves https://github.com/open-telemetry/opentelemetry-proto/issues/147.

james-bebbington commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,209 @@ type (
+				err = aggregation.ErrNoSubtraction
+			}
+		} else {
+			err = value.cumulative.Merge(value.current, key.descriptor)
```
Above we explicitly do a subtraction. Here we do a `Merge` which iiuc because of the conditions to get to this point, will always have the effect of performing an addition?

If that's correct, perhaps we could make the addition here explicit so that's its consistent with the above code and really clear that we are just converting back and forth between cumulatives & deltas? (I might be totally wrong)

Also maybe consistent comments would help:

Above: `// delta value = current cumulative value - previous cumulative value`
Here: `// cumulative value = previous cumulative value + current delta value`

jmacd replied:
I've added your recommended comments.

We could consider renaming the `Merge()` function to `Add()`, but there are cases where that may create confusion. There is (maybe) a fine distinction here: Aggregator's aren't required to implement Subtract, they should only do so when they deal in "Additive" measurements.  As it stands, "Merge" means Add for the Adding instruments and it means Combine for the Grouping instruments.

jmacd replied:
Subtract is only used in the case where an instrument and exporter use PrecomputedSum **AND** is stateful because of a delta exporter. Subtract will be needed by default, only for a (dog)statsd exporter with (UpDown)SumObserver instruments.

Merge is the ordinary case, handling all the other possibilities including both Adding and Grouping instruments. 

I added support for Sum.Subtract in this PR.  If we wanted to compute a distribution of SumObserver values (as opposed to a simple sum), we'd choose a stateful Histogram aggregator which can also support Subtract (however I did not add it in this PR).

james-bebbington commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,209 @@ type (
+	sameRound := b.state.finishedCollection == value.updated
+	value.updated = b.state.finishedCollection
+
+	// An existing record will be found when:
```
I don't really get this comment. It seems like an existing record will always be found on the second round since we always populate this map if its not set above. 

Is this supposed to say "An existing record will be found **in the same round** when:"? But then why would (a) effect this?

MrAlias replied:
I think this is confusing as well.

I think this is describing when an "existing record will _not_ be found in the same round" and `sameRound` will be false.

I think it would be helpful to be consistent in terminology here. Instead of `record` -> `value`, and instead of the term `round` maybe something more descriptive of the value's state. Maybe `continuous`?

james-bebbington commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,209 @@ type (
+		value.current = agg
+		return nil
+	}
+	if value.delta == nil {
```
I'm attempting to review this PR with limited knowledge of the SDK pipeline, so it may just be I'm lacking context, but I find it very hard to figure out what's happening in this function.

I think part of the difficulty for me is distinguishing between work that's done to convert between cumulative & delta values, and work that's done to aggregate for other reasons (when multiple accumulators are used?). In particular, I thought the `delta` aggregation was only needed if we had a pre-computed sum that needs to be converted back to a delta for the exporter (as per line 138), so I'm not sure why we are using it here.

Is it possible to separate this logic more clearly? I'd try to make a suggestion of how to go about that, but not sure I'm following along well enough.

MrAlias replied:
I agree, this could be clearer.

I feel like breaking apart to separate functions could help. That, or really explicit comments targeted at devs with no knowledge of the pipeline being added.

james-bebbington replied:
Thanks for adding comments. It's definitely easier to understand now.

I still think some code changes could improve readability as well though. In particular, do we have to use the `delta` aggregator property for both storing the conversion from `cumulative` to `delta` AND for storing the multi-aggregator temp results? In fact, can we not just store the multi-aggregator temp result in `current` & use a flag to determine if we have performed the Clone? (which would happen when we see the second accumulator in the same collection)

jmacd replied:
You are right, we do not have to use the `delta` aggregator in two different ways.  If we use `current` and a flag, it won't be possible to keep the state that was allocated for storing `delta` values (in both cases). The way this is coded allows referring to an Aggregator owned by the Accumulator in the common case of a single Accumulator, as well as the less-common case of a stateful Asynchronous instrument.

I will also say that in #799 I tried doing this in another way, and it was even worse. I'm sure this can be improved, and I want to assure you that you seem to understand what's happening very well!

jmacd commented on api/metric/number.go:
```diff
@@ -646,3 +646,15 @@ func (n *Number) compareWithZero(kind NumberKind) int {
+
+// - sign change
+
+func ChangeSign(kind NumberKind, nn Number) Number {
```
It could be changed to `func (n *Number) ChangeSign(kind NumberKind)`.  It would be confusing to take a pointer receiver and not modify the receiver, so there would be no return value.

However, every use of this function (and every test) was creating a new value, so changing it to modify its receiver would mean lots of change in the code for little value, so I thought it would be better to name this like a constructor, `NewNumberSignChange(Kind, Number) Number`. Does this help?

jmacd commented on api/metric/number.go:
```diff
@@ -646,3 +646,15 @@ func (n *Number) compareWithZero(kind NumberKind) int {
+
+func ChangeSign(kind NumberKind, nn Number) Number {
+	switch kind {
+	case Int64NumberKind:
```
Right, the result isn't well defined.  I don't think we need a Uint64 case unless we were to add support for Uint64 instruments. The only uses of Uint64 support were atomic counts, which we no longer use. I added a TODO.

jmacd commented on exporters/metric/prometheus/prometheus.go:
```diff
@@ -173,6 +162,14 @@ func (e *Exporter) Controller() *pull.Controller {
 
+func (e *Exporter) ExportKindFor(*metric.Descriptor, aggregation.Kind) export.ExportKind {
+	// NOTE: Summary values should use Delta aggregation, then be
+	// combined into a sliding window, see the TODO below.
```
Thanks. I downgraded the TODO on the following line to a NOTE because I don't have enough information (and the public documents on GaugeDelta are sketchy at best).

jmacd commented on sdk/metric/controller/push/push.go:
```diff
@@ -60,7 +60,7 @@ func New(selector export.AggregationSelector, exporter export.Exporter, opts ...
 		c.Timeout = c.Period
 	}
 
-	integrator := simple.New(selector, c.Stateful)
```
Fixed.

jmacd commented on sdk/export/metric/metric.go:
```diff
@@ -292,3 +324,52 @@ func (r Record) StartTime() time.Time {
+	return kind&has != 0
+}
+
+// ExportKindFor returns a constant, as an implementation of ExportKindFor.
```
Fixed.

jmacd commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -29,31 +29,47 @@ import (
-		values map[batchKey]batchValue
+		values map[stateKey]*stateValue
 
 		// Note: the timestamp logic currently assumes all
 		// exports are deltas.
 
```
😀

jmacd commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,209 @@ type (
 }
 
-func (b *batch) ForEach(f func(export.Record) error) error {
+// ForEach iterates through the CheckpointSet, for passing an
```
Fixed.

jmacd commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,209 @@ type (
+				agg = value.current.Aggregation()
+			}
+			start = b.intervalStart
+		}
```
Fixed and tested.

jmacd commented:
@james-bebbington and @MrAlias I think I've addressed your concerns, but you should be the judge. Thanks.

james-bebbington commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,271 @@ type (
+	sameCollection := b.state.finishedCollection == value.updated
+	value.updated = b.state.finishedCollection
+
+	// An existing value will be found for some stateKey when:
```
You changed the terminology to be a bit clearer, but I still think this comment is wrong. We set `b.state.values[key] = newValue` on line 146 outside of the `if stateful { ... }` block, and indeed I believe we need to otherwise `value.current` would never be set (e.g. line 184).

Should this not just be "sameCollection will be true when: (b) ..."

Case (a) seems to be redundant for the code that follows (which to me seems to only be concerned with multi-accumulator aggregation).

jmacd replied:
I can see how the comment is confusing. I meant "An existing value will be found _AT THIS POINT IN THE CODE_". To reach this point in the code, we have to have already passed through line 146 (in the current collection, if stateless, or in any previous collection, if not stateless). I added a point that (a) can be true if stateless, and (b) is true the second time through whether stateful or not. I'm struggling to improve on this...

jmacd commented on sdk/metric/integrator/simple/simple.go:
```diff
@@ -68,97 +84,271 @@ type (
+		var err error
+		if mkind.PrecomputedSum() {
+			// delta_value = current_cumulative_value - previous_cumulative_value
+			if subt, ok := value.current.(export.Subtractor); ok {
+				err = subt.Subtract(value.cumulative, value.delta, key.descriptor)
```
I did this and added a bit more comment. I considered what to name such an in-place operation, and the exercise left me feeling like it would add more confusion than it would relieve. There's not a good way to say "T <- S - T".  I'd be happy to do this in a future PR.

james-bebbington replied:
Or even better would it be possible to just return the result so this could be written as:
```go
value.delta = currentSubtractor.Subtract(value.cumulative, key.descriptor)
```
(I seem to remember thinking that's not possible but can't remember why)

jmacd commented:
Thanks @MrAlias and @james-bebbington!

jmacd commented:
I failed to push my final commits in response to @james-bebbington. Sorry! I will roll them into a separate PR.

jmacd commented:
(it was this chunk)

```
diff --git a/sdk/metric/integrator/simple/simple.go b/sdk/metric/integrator/simple/simple.go
index 40ccce33..def51169 100644
--- a/sdk/metric/integrator/simple/simple.go
+++ b/sdk/metric/integrator/simple/simple.go
@@ -151,9 +151,20 @@ func (b *Integrator) Process(accum export.Accumulation) error {
 	sameCollection := b.state.finishedCollection == value.updated
 	value.updated = b.state.finishedCollection
 
-	// An existing value will be found for some stateKey when:
-	// (a) stateful aggregation is being used
-	// (b) multiple accumulators are being used.
+	// At this point in the code, we have located an existing
+	// value for some stateKey.  This can be because:
+	//
+	// (a) stateful aggregation is being used, the entry was
+	// entered during a prior collection, and this is the first
+	// time processing an accumulation for this stateKey in the
+	// current collection.  Since this is the first time
+	// processing an accumulation for this stateKey during this
+	// collection, we don't know yet whether there are multiple
+	// accumulators at work.  If there are multiple accumulators,
+	// they'll hit case (b) the second time through.
+	//
+	// (b) multiple accumulators are being used, whether stateful
+	// or not.
 	//
 	// Case (a) occurs when the instrument and the exporter
 	// require memory to work correctly, either because the
@@ -273,9 +284,10 @@ func (b *Integrator) FinishCollection() error {
 		// delta or a cumulative aggregation.
 		var err error
 		if mkind.PrecomputedSum() {
-			// delta_value = current_cumulative_value - previous_cumulative_value
-			if subt, ok := value.current.(export.Subtractor); ok {
-				err = subt.Subtract(value.cumulative, value.delta, key.descriptor)
+			if currentSubtractor, ok := value.current.(export.Subtractor); ok {
+				// This line is equivalent to:
+				// value.delta = currentSubtractor - value.cumulative
+				err = currentSubtractor.Subtract(value.cumulative, value.delta, key.descriptor)
 
 				if err == nil {
 					err = value.current.SynchronizedCopy(value.cumulative, key.descriptor)
@@ -284,7 +296,8 @@ func (b *Integrator) FinishCollection() error {
 				err = aggregation.ErrNoSubtraction
 			}
 		} else {
-			// cumulative_value = previous_cumulative_value + current_delta_value
+			// This line is equivalent to:
+			// value.cumulative = value.cumulative + value.delta
 			err = value.cumulative.Merge(value.current, key.descriptor)
 		}
 		if err != nil {
```
